### ISSUE TITLE
Strange bug with Strings

### ISSUE THREAD

**Issue by slaff** [SmingHub/Sming#1451]
@mikee47 Strangely enough the following code produces incorrect result:

```
http_method_str(request->method) + String(" ") + request->uri.getPathWithQuery() + " HTTP/1.1\r\n"
```

Instead of the complete HTTP request line I am getting just "HTTP/1.1\r\n". The issue is happening after your recent PRs were merged (related to the String class). Can you check if you can reproduce it or it is just me and my compiler with `mforce-l32`?

**Comment by slaff** [comment]
I narrowed down the problem to the following code:
```
        String path = "/path/to";
	String query;

	if(path.length() + query.length() > 0) {
		String tmpString = path + query;
		debugf("[%s]", tmpString.c_str());
	}
```

Expected output: [/path/to].
Actual output: [(null)]

Which is not correct and might break the code in multiple places.
@mikee47 when you have time please submit a fix for this issue.

**Comment by mikee47** [comment]
@slaff That is weird. I tested it with:
```
String content = http_method_str(request.method) + String(" ") + request.uri.pathWithQuery() + " HTTP/1.1\n";
debug_i("request: %s", content.c_str());
```

And it's fine.

**Comment by mikee47** [comment]
Fix is in WString.cpp, swap lines 279 & 280 to read:
```
  if (length == 0) return true;
  if (!cstr) return false;
```

I'll review the rest of it see if there are any similar conditions then submit a PR.

**Comment by mikee47** [comment]
diff --git a/Sming/Wiring/WString.cpp b/Sming/Wiring/WString.cpp
index 6a87294f..c6e5e873 100644
--- a/Sming/Wiring/WString.cpp
+++ b/Sming/Wiring/WString.cpp
@@ -276,8 +276,8 @@ bool String::concat(const String &s)
 bool String::concat(const char *cstr, unsigned int length)
 {
   unsigned int newlen = len + length;
-  if (!cstr) return false;
   if (length == 0) return true;
+  if (!cstr) return false;
   if (!reserve(newlen)) return false;
   memmove(buffer + len, cstr, length);
   buffer[newlen] = '\0';

**Comment by mikee47** [comment]
Another thing, which may be of concern. As an example, consider:
```
String tmp = String("test ") + String(54) + String(" again") + anotherString;
```
If any of the allocations in the expression fail, then the result, tmp, should be an invalid string. However, with the changes I made this will not happen and the result will be unpredictable.

Thoughts?

**Comment by slaff** [comment]
> result will be unpredictable.

Unpredictable results is bad and is completely negating any good results of not setting initially the string to be "" instead of nullptr.  Especially for a class that is one of the most used core classes.

Mike can you make another PR, alongside the current PR #1452, where you change the code so that a new String object has always "" instead of nullptr + change accordingly concat and the other methods.

**Comment by mikee47** [comment]
That's unpredictable in the event of heap allocation failure - #969 addresses this in a wider context.
With the original unmodified version of String, the result would be empty strings being used, which I guess is a little more predictable, though still undesirable, but still requires the caller to check that a String is valid before using it; I cannot recall seeing that done anywhere.

@slaff I could do with some steering on how Sming should deal with heap allocation failures.

**Comment by slaff** [comment]
@mikee47 Can you rephrase and add 

`If any of the allocations in the expression fail, then the result, tmp, should be an invalid string. However, with the changes I made this will not happen and the result will be unpredictable.` 

as comment in the code of PR #1452 that mentions this unlikely-but-possible situation.
